/***************************************************************************
 *   rshtdlg.cpp                                                           *
 *   Copyright (C) 2014 by Daniel Suson                                    *
 *   suson@calumet.purdue.edu                                              *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 *                                                                         *
 *   In addition, as a special exception, the copyright holders give       *
 *   permission to link the code of this program with any edition of       *
 *   the Qt library by Trolltech AS, Norway (or with modified versions     *
 *   of Qt that use the same license as Qt), and distribute linked         *
 *   combinations including the two.  You must obey the GNU General        *
 *   Public License in all respects for all of the code used other than    *
 *   Qt.  If you modify this file, you may extend this exception to        *
 *   your version of the file, but you are not obligated to do so.  If     *
 *   you do not wish to do so, delete this exception statement from        *
 *   your version.                                                         *
 ***************************************************************************
 * Description:                                                            *
 *                                                                         *
 * This class defines the rsht dialog structure associated with            *
 * the Qt4 user interface.                                                 *
 * ----------------------- Change Log ------------------------------------ *
 * Date    Name         Description                                        *
 *                                                                         *
 ***************************************************************************
 * This software evolved from work that was done by Eric Hivon at the      *
 * California Institute of Technology on the MASTER algorithm (Hivon, E.,  *
 * et al., ApJ, vol 567, p 2).  It adapts or interfaces with these         *
 * packages:                                                               *
 *                                                                         *
 * HealPix         (http://www.eso.org/science/healpix/)                   *
 * FTOOLS & FITSIO (http://heasarc.gsfc.nasa.gov/docs/corp/software.html)  *
 *                                                                         *
 * Any publications or presentations that use results generated by this    *
 * program must reference the fact that it was developed by Daniel Suson   *
 ***************************************************************************/
#include "rshtdlg.h"
#include "ui_rshtdlg.h"

rshtDialog::rshtDialog(association *assoc) :
            ui(new Ui::rshtDialog) {
  /* set up the user interface first */
  ui->setupUi(this);

  dataAssoc = assoc;

  ui->minBox->setValue(0);
  ui->maxBox->setValue(1);
  ui->iterationBox->setValue(1);
  
  minIndex = 0;
  maxIndex = 1;
  noIterations = 1;
    
  /* set up signals and slots */
  connect(ui->buttonBox, &QDialogButtonBox::accepted, this, &rshtDialog::finalize);
  connect(ui->buttonBox, &QDialogButtonBox::rejected, this, &rshtDialog::cancel);
  connect(ui->buttonBox, &QDialogButtonBox::helpRequested, this, &rshtDialog::help);
  
  /* initialize internal variables*/
//  currentHandle = -1;
//  dataHandle = -1;
  dirty = false;
  isConfigured = false;
  
  /* set initially displayed values */
  ui->minBox->setValue(minIndex);
  ui->maxBox->setValue(maxIndex);
  ui->iterationBox->setValue(noIterations);

}

rshtDialog::~rshtDialog() {
  delete ui;
}

void rshtDialog::configure() {
//  QString title, message;
  
  dirty = false;
//  dataHandle = analyzer_get_instance();
//  if (dataHandle < 0) {
//    title = QString(tr("Error code returned"));
//    message = QString(tr("No data chain found:\nError code is %1 - %2")).arg(dataHandle).arg(analyzer_get_error(dataHandle));
//    QMessageBox::critical(this,title,message);
//    return;
//  }

  if (isConfigured) {
    ui->minBox->setValue(minIndex);
    ui->maxBox->setValue(maxIndex);
    ui->iterationBox->setValue(noIterations);
  }

  exec();
}

void rshtDialog::reset() {
  ui->minBox->setValue(0);
  ui->maxBox->setValue(1);
  ui->iterationBox->setValue(1);
  minIndex = 0;
  maxIndex = 1;
  noIterations = 1;
  dirty = false;
  isConfigured = false;
  
//  while (currentHandle >= 0) {
//    analyzer_transformer_release(currentHandle);
//    currentHandle--;
//  }
}

void rshtDialog::validate() {
  dirty = false;  // assume nothing has changed
  
  /* determine minimum and maximum indices */
  int oldMin = minIndex;
  int oldMax = maxIndex;
  minIndex = ui->minBox->value();
  if (oldMin != minIndex)
    dirty = true;
  maxIndex = ui->maxBox->value();
  if (oldMax != maxIndex)
    dirty = true;

  /* determine iterations */
  int oldNoIterations = noIterations;
  noIterations = ui->iterationBox->value();
  if (oldNoIterations != noIterations)
    dirty = true;
  
  isConfigured = true;
  return;
}

/*
int rshtDialog::transform() {
  analyzer_transform_hdl th = analyzer_get_transformer();
  analyzer_hdl hdl = analyzer_get_instance();
  
  // Set transformer operational parameters
  analyzer_trans_set_spectral_indices(th, minIndex, maxIndex);
  analyzer_trans_set_iterations(th, iterations);
  
  // transform available data types
  analyzer_data_t pixType = ANALYZER_PIXEL;
  analyzer_data_t transType = ANALYZER_TRANSFORM;
  int type = static_cast<int>(pixType);
  int offset = 14, result = 0;
  
  while (type) {
    if (analyzer_check_existence(hdl,pixType) && !analyzer_check_existence(hdl,transType)) {
      result = analyzer_transform(th,hdl,pixType);
      if (result < 0)
        return result;
    }
    pixType = static_cast<analyzer_data_t>(++type);
    transType = static_cast<analyzer_data_t>(type + offset);
    if (pixType == ANALYZER_PIXEL_OCCUPANCY)
      break;
  }
  
  return result;
}

int rshtDialog::configureInterface() {
  QString title, message;

  // check to see if a transformer already exists
  if (currentHandle >= 0) {
    QMessageBox::StandardButton reply;
    title = QString(tr("Replace current transformer?"));
    message = QString(tr("A transformer currently exists.\nDo you want to replace it?"));
    reply = QMessageBox::question(this,title,message,QMessageBox::Yes | QMessageBox::No,QMessageBox::No);
    if (reply == QMessageBox::Yes)
      analyzer_transformer_release(currentHandle);
  }
  
  // create a new transformer either way
  currentHandle = analyzer_create_transformer(ANALYZER_TRANS_RSHT);
  if (currentHandle < 0) {
    title = QString(tr("Error code returned"));
    message = QString(tr("The attempt to create a new transformer failed.\nError code is %1 - %2")).arg(currentHandle).arg(analyzer_get_error(currentHandle));
    QMessageBox::critical(this,title,message);
    return currentHandle;
  }
  
  return 0;
}
*/

void rshtDialog::finalize() {
  validate(); //check and store selections
  dataAssoc->transformationEngineType(Rsht);

  Q_EMIT transformData();
  accept();
}

void rshtDialog::help() {

}

void rshtDialog::cancel() {
  QString title, message;
//  QMessageBox::StandardButton reply;

  if (dirty) {
    title = QString(tr("Transformation parameters will be reset"));
    message = QString(tr("You have chosen to cancel after changing some parameters. These will be reset.\nThis will not affect any parameters stored in the transformation engine"));
    QMessageBox::information(this,title,message);
    reset();
    dirty = false;
  }
  Q_EMIT transformationCancelled();
  close();
}
